src/main/java/org/jvnet/hudson/test/RealJenkinsRule.java

```diff
@@ -608,25 +618,42 @@ private static boolean supportsPortFileName(String war) throws IOException {
 
     private static int readPort(File portFile) throws IOException {
         String s = FileUtils.readFileToString(portFile, StandardCharsets.UTF_8);
+        if (StringUtils.isEmpty(s)) {
+            LOGGER.log(Level.WARNING, () -> String.format("PortFile: %s exists, but value is still not written", portFile.getAbsolutePath()));
+            return 0;
+        }
         try {
             return Integer.parseInt(s);
         } catch (NumberFormatException e) {
             throw new AssertionError("Unable to parse port from " + s + ". Jenkins did not start.");
         }
     }
 
+    /**
+     * Stops Jenkins and releases any system resources associated
+     * with it. If Jenkins is already stopped then invoking this
+     * method has no effect.
+     */
     public void stopJenkins() throws Throwable {
-        endpoint("exit").openStream().close();
-        if (!proc.waitFor(60, TimeUnit.SECONDS) ) {
-            System.err.println("Jenkins failed to stop within 60 seconds, attempting to kill the Jenkins process");
-            proc.destroyForcibly();
-            throw new AssertionError("Jenkins failed to terminate within 60 seconds");
+        if (!stopping.get()) {
+            stopping.set(true);
```

Comment:
If you are not actually using atomic swap operations, why is this field not simply a `boolean`?

Reply:
oh yes, done in [c3cfed6](https://github.com/jenkinsci/jenkins-test-harness/pull/465/commits/c3cfed66853ce70ca4403094fb67b15bf1760958)